# Lab | Data Structuring and Combining Data

## Challenge 1: Combining & Cleaning Data

In this challenge, we will be working with the customer data from an insurance company, as we did in the two previous labs. The data can be found here:
- https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv

But this time, we got new data, which can be found in the following 2 CSV files located at the links below.

- https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file2.csv
- https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file3.csv

Note that you'll need to clean and format the new data.

Observation:
- One option is to first combine the three datasets and then apply the cleaning function to the new combined dataset
- Another option would be to read the clean file you saved in the previous lab, and just clean the two new files and concatenate the three clean datasets

In [19]:
import pandas as pd

Functions and cleanup

In [20]:
def clean_columns(dataframe):
    dataframe.columns = [col.lower().replace(' ','_') if col != 'ST' else 'state' for col in dataframe.columns ]

In [21]:
def clean_inconsistent_gender_values(value):
    val = str(value)
    if len(val) >= 1:
        if val.startswith('f') or val.startswith('F'):
            correct_value = "F"
            return correct_value
        elif val.startswith('m') or val.startswith('M'):
            correct_value = "M"
            return correct_value
        else:
            return 'NULL'

In [22]:
def clean_inconsistent_state_values(value):
    val = str(value)
    if len(val) == 2:
        if val == 'AZ':
            correct_value = "Arizona"
            return correct_value
        elif val == 'WA':
            correct_value = "Washington"
            return correct_value
        else:
            return val
    elif len(val) > 2 and len(val) < 5:
        if val == 'Cali':
            correct_value = 'California'
        else:
            return 'NULL'
    else:
        return val

In [23]:
def clean_education(value):
    val = str(value)
    if val == 'Bachelors':
        return 'Bachelor'
    else:
        return val

In [24]:
def clean_customer_lifetime_vale(value):
    val = str(value)
    if val[-1] == '%':
        val = val.replace('%','')
        return float(val)
    else:
        try:
            return float(val)
        except:
            return 0

In [25]:
def clean_complaints(value):
    val = str(value).split('/')
    if len(val) > 1:
        try: 
            res = int(val[1])
            return res
        except:
            print(f'{val} not valid int')
    else:
        try:
            res = int(val[0])
            return res
        except:
            return None

In [26]:
df = pd.read_csv('https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv')
df2 = pd.read_csv('https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file2.csv')
df3 = pd.read_csv('https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file3.csv')

In [28]:
def clean_df(dataframe):
    """
    Cleans 'gender', 'education', 'customer_lifetime_vale', 'number_of_open_compaints' columns from dataframe parameter
    """
    clean_columns(dataframe)
    dataframe['gender'] = dataframe['gender'].apply(clean_inconsistent_gender_values)
    dataframe['state'] = dataframe['state'].apply(clean_inconsistent_state_values)
    dataframe['education'] = dataframe['education'].apply(clean_education)
    dataframe['customer_lifetime_value'] = dataframe['customer_lifetime_value'].apply(clean_customer_lifetime_vale)
    dataframe['number_of_open_complaints'] = dataframe['number_of_open_complaints'].apply(clean_complaints)

In [29]:
#DataFrame 1
clean_df(df)

In [30]:
#DataFrame 2
clean_df(df2)

In [31]:
#DataFrame 3
clean_df(df3)

In [35]:
display(df.head(5))

customer       state gender             education  customer_lifetime_value  \
0  RB50392  Washington   NULL                Master                      NaN   
1  QZ44356     Arizona      F              Bachelor                697953.59   
2  AI49188      Nevada      F              Bachelor               1288743.17   
3  WW63253  California      M              Bachelor                764586.18   
4  GA49547  Washington      M  High School or Below                536307.65   

    income  monthly_premium_auto  number_of_open_complaints     policy_type  \
0      0.0                1000.0                        0.0   Personal Auto   
1      0.0                  94.0                        0.0   Personal Auto   
2  48767.0                 108.0                        0.0   Personal Auto   
3      0.0                 106.0                        0.0  Corporate Auto   
4  36357.0                  68.0                        0.0   Personal Auto   

   vehicle_class  total_claim_amount  
0  Four-Door Car            2.704934  
1  Four-Door Car         1131.464935  
2   Two-Door Car          566.472247  
3            SUV          529.881344  
4  Four-Door Car           17.269323

In [36]:
display(df2.head(5))

customer       state gender education  customer_lifetime_value  income  \
0  GS98873     Arizona      F  Bachelor                323912.47   16061   
1  CW49887  California      F    Master                462680.11   79487   
2  MY31220  California      F   College                899704.02   54230   
3  UH35128      Oregon      F   College               2580706.30   71210   
4  WH52799     Arizona      F   College                380812.21   94903   

   monthly_premium_auto  number_of_open_complaints  total_claim_amount  \
0                    88                          0               633.6   
1                   114                          0               547.2   
2                   112                          0               537.6   
3                   214                          1              1027.2   
4                    94                          0               451.2   

      policy_type  vehicle_class  
0   Personal Auto  Four-Door Car  
1    Special Auto            SUV  
2   Personal Auto   Two-Door Car  
3   Personal Auto     Luxury Car  
4  Corporate Auto   Two-Door Car

In [37]:
display(df3.head(5))

customer       state  customer_lifetime_value             education gender  \
0  SA25987  Washington              3479.137523  High School or Below      M   
1  TB86706     Arizona              2502.637401                Master      M   
2  ZL73902      Nevada              3265.156348              Bachelor      F   
3  KX23516  California              4455.843406  High School or Below      F   
4  FN77294  California              7704.958480  High School or Below      M   

   income  monthly_premium_auto  number_of_open_complaints    policy_type  \
0       0                   104                          0  Personal Auto   
1       0                    66                          0  Personal Auto   
2   25820                    82                          0  Personal Auto   
3       0                   121                          0  Personal Auto   
4   30366                   101                          2  Personal Auto   

   total_claim_amount  vehicle_class  
0          499.200000   Two-Door Car  
1            3.468912   Two-Door Car  
2          393.600000  Four-Door Car  
3          699.615192            SUV  
4          484.800000            SUV

In [43]:
full_df = pd.concat([df, df2, df3], axis=0)
full_df.shape

(12074, 11)

In [44]:
def clean_null_dup_values(dataframe):
    dataframe.dropna(inplace=True)
    dataframe.drop_duplicates(inplace=True)

In [45]:
clean_null_dup_values(full_df)

In [46]:
full_df.shape

(9007, 11)

# Challenge 2: Structuring Data

In this challenge, we will continue to work with customer data from an insurance company, but we will use a dataset with more columns, called marketing_customer_analysis.csv, which can be found at the following link:

https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis_clean.csv

This dataset contains information such as customer demographics, policy details, vehicle information, and the customer's response to the last marketing campaign. Our goal is to explore and analyze this data by performing data cleaning, formatting, and structuring.

In [61]:
marketing_df = pd.read_csv('https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis_clean.csv')
marketing_df.shape

(10910, 27)

In [62]:
marketing_df.columns

Index(['unnamed:_0', 'customer', 'state', 'customer_lifetime_value',
       'response', 'coverage', 'education', 'effective_to_date',
       'employmentstatus', 'gender', 'income', 'location_code',
       'marital_status', 'monthly_premium_auto', 'months_since_last_claim',
       'months_since_policy_inception', 'number_of_open_complaints',
       'number_of_policies', 'policy_type', 'policy', 'renew_offer_type',
       'sales_channel', 'total_claim_amount', 'vehicle_class', 'vehicle_size',
       'vehicle_type', 'month'],
      dtype='object')

In [63]:
clean_df(marketing_df)

In [78]:
marketing_df.head(1)

unnamed:_0 customer    state  customer_lifetime_value response coverage  \
0           0  DK49336  Arizona               4809.21696       No    Basic   

  education effective_to_date employmentstatus gender  ...  \
0   College        2011-02-18         Employed      M  ...   

   number_of_policies     policy_type        policy  renew_offer_type  \
0                   9  Corporate Auto  Corporate L3            Offer3   

   sales_channel  total_claim_amount  vehicle_class  vehicle_size  \
0          Agent               292.8  Four-Door Car       Medsize   

  vehicle_type month  
0            A     2  

[1 rows x 27 columns]

1. You work at the marketing department and you want to know which sales channel brought the most sales in terms of total revenue. Using pivot, create a summary table showing the total revenue for each sales channel (branch, call center, web, and mail).
Round the total revenue to 2 decimal points.  Analyze the resulting table to draw insights.

In [110]:
pivot_table = marketing_df.pivot_table(index = 'sales_channel', columns= 'state', values='customer_lifetime_value', aggfunc=('sum'))
pivot_table['total_revenue'] = pivot_table.sum(axis=1)
pivot_table =  pivot_table.round(2) 
display(pivot_table)

state             Arizona   California      Nevada      Oregon  Washington  \
sales_channel                                                                
Agent          5761629.34  12464967.45  3415783.66  8760327.92  2655179.48   
Branch         3989670.78   9539065.86  2034128.25  6630912.94  2165423.37   
Call Center    3282468.58   6440770.72  1629196.40  4730256.64  1281596.03   
Web            2306010.67   4825795.69  1130450.29  3575850.98   859525.27   

state          total_revenue  
sales_channel                 
Agent            33057887.85  
Branch           24359201.21  
Call Center      17364288.37  
Web              12697632.90

Agents are the best sales_channel with 33057887.85 total revenue, followed by Branch with 24359201.21, Call Center with 17364288.37 and lastly Web with 12697632.90

2. Create a pivot table that shows the average customer lifetime value per gender and education level. Analyze the resulting table to draw insights.

In [111]:
pivot_table2 = marketing_df.pivot_table(index = 'gender', columns= 'education', values='customer_lifetime_value', aggfunc=('mean'))
pivot_table2['total_revenue'] = pivot_table2.sum(axis=1)
pivot_table2 =  pivot_table2.round(2) 
display(pivot_table2)

education  Bachelor  College   Doctor  High School or Below   Master  \
gender                                                                 
F           7874.27  7748.82  7328.51               8675.22  8157.05   
M           7703.60  8052.46  7415.33               8149.69  8168.83   

education  total_revenue  
gender                    
F               39783.88  
M               39489.92

We can relate higher education with lower total revenue except for 'Master' which is number 2
Females generate 3000 more total revenue, with the highest education in terms of revenue being High School or bellow.
Males with college, masters and doctorates generate more total revenue than females in college, however in general the total revenue is slighly lower.

## Bonus

You work at the customer service department and you want to know which months had the highest number of complaints by policy type category. Create a summary table showing the number of complaints by policy type and month.
Show it in a long format table.

*In data analysis, a long format table is a way of structuring data in which each observation or measurement is stored in a separate row of the table. The key characteristic of a long format table is that each column represents a single variable, and each row represents a single observation of that variable.*

*More information about long and wide format tables here: https://www.statology.org/long-vs-wide-data/*

In [ ]:
# Your code goes here